### PR TITLE
added core::mem::reshape

### PR DESCRIPTION
**Pull request by hazelwiss**
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/130849)*

tracking issue: https://github.com/rust-lang/rust/issues/130848

### CODE REVIEW

File: library/core/src/mem/mod.rs [hunk 1]
```diff
@@ -874,6 +874,38 @@ pub const fn replace<T>(dest: &mut T, src: T) -> T {
...
+/// ```
+#[inline]
+#[unstable(feature = "mem_reshape", issue = "130848")]
+pub fn reshape<T>(dest: &mut T, f: impl FnOnce(T) -> T) {
+    // SAFETY: We read from `dest` but directly write to it afterwards,
+    // such that the old value is not duplicated. Nothing is dropped and
+    // nothing here can panic.
+    unsafe {
+        let result = ptr::read(dest);
+        ptr::write(dest, f(result));
```
Comment (Noratrieb) [inline library/core/src/mem/mod.rs:905]: this is not sound if `f` unwinds, as the value will be dropped twice. see https://docs.rs/replace_with/latest/replace_with/ for details. the design space here is nontrivial, as there are many possible behaviors to do on unwind. this should at least go through an [ACP](https://std-dev-guide.rust-lang.org/development/feature-lifecycle.html), but probably better for more general discussion on internals.rust-lang.org. i'd also suggesting searching there for previous discussion about "replace_with", which is what this function is generally called

Comment (hazelwiss) [reply library/core/src/mem/mod.rs:905]: yeah I just detected that myself and ran into a wall. Sounds good :+1:

Comment (hazelwiss) [reply library/core/src/mem/mod.rs:905]: has there been any discussion about a feature to ensure a closure does not panic?

Comment (Noratrieb) [reply library/core/src/mem/mod.rs:905]: hm. this has probably come up in vague ways in many places but nothing concrete.

Comment (Noratrieb) [reply library/core/src/mem/mod.rs:905]: either way it's definitely both that will arrive any time soon, so it probably makes sense to close the PR and tracking issue for now

Comment (Noratrieb) [reply library/core/src/mem/mod.rs:905]: thank you for it anyways :3

Comment (hazelwiss) [reply library/core/src/mem/mod.rs:905]: :+1:

File: library/core/src/mem/mod.rs [hunk 2]
```diff
@@ -874,6 +874,38 @@ pub const fn replace<T>(dest: &mut T, src: T) -> T {
...
+/// # Examples
+///
+/// A simple example:
+///
+/// ```
+/// use std::mem;
+///
+/// let mut v: Vec<i32> = vec![1, 2];
+///
+/// mem::replace(&mut v, |mut v| { v.push(3); v });
```
Comment (narpfel) [inline library/core/src/mem/mod.rs:894]:
```suggestion
/// mem::reshape(&mut v, |mut v| { v.push(3); v });
```
Should this be `reshape`?

File: library/core/src/mem/mod.rs (hunk 1 above)
Comment (juntyr) [reply library/core/src/mem/mod.rs:905]: Could you execute the closure inside an extern "C" wrapper func that would abort on unwind? I think I recently saw a PR on a helper that exposed abort on unwind (mirroring the catch_unwind function) using that technique

Comment (workingjubilee) [reply library/core/src/mem/mod.rs:905]: You're thinking of https://github.com/rust-lang/rust/pull/130339

File: library/core/src/mem/mod.rs [hunk 3]
```diff
@@ -874,6 +874,38 @@ pub const fn replace<T>(dest: &mut T, src: T) -> T {
...
+/// assert_eq!(vec![1, 2, 3], v);
+/// ```
+#[inline]
+#[unstable(feature = "mem_reshape", issue = "130848")]
+pub fn reshape<T>(dest: &mut T, f: impl FnOnce(T) -> T) {
```
Comment (bjorn3) [inline library/core/src/mem/mod.rs:899]: I don't think this is a good name. When I first came across the tracking issue in the issue list, I though this was something transmute related.

**Comment by Dylan-DPC** [comment]
@hazelwiss do you still want to go ahead with this? this needs an ACP which hasn't been return and i think above some folks aren't happy with the naming and other things

**Comment by workingjubilee** [comment]
This needs a better motivation that is not easily replaced by the mutable reference.

**Comment by workingjubilee** [comment]
It also probably needs to be named something similar to [`replace_with`](https://docs.rs/replace_with/latest/replace_with/), because that's what this is.

**Comment by clarfonthey** [comment]
I was under the impression that there were concerns that this kind of function cannot be sound, and that's why it isn't offered, but I could be wrong. The main problem is that you have to ensure that you can never observe the value while it's been moved out of the reference, which requires ensuring that the closure either cannot panic or cannot return after panicking, i.e. aborts on panic.

**Comment by workingjubilee** [comment]
Yes, this is [RFC 1736](https://github.com/rust-lang/rfcs/pull/1736) all over again.

**Comment by workingjubilee** [comment]
> @Dylan-DPC Yes! I was **under the impression** people wanted to close this PR in favour of something else, but I would gladly go ahead with this. I can fix both the naming and look into making an ACP for this.

What I want is to either
1. not close this PR until I am sure you have been given the best opportunity to understand why it will be closed, as I'd rather you not leave with an *impression*
2. see this undergo the changes needed to make it something that will not be rejected, which will probably not be a minor edit here and there

Which path you wish to go down is your choice.

You did not respond to the recommendation to close this issue, so I was not certain you had read the latest comments, and I mostly focused on advancing the other PRs in my queue.

**Comment by hazelwiss** [comment]
@workingjubilee I don't have a problem closing this issue/pr especially if it's a duplicate! Whichever you guys think is the best decision! I have no particular impression, all the feedback makes sense.

**Comment by workingjubilee** [comment]
I don't believe this is what I would call a duplicate. A duplicate implies that the previous issue fully encompasses the current one, but it has been some time since `replace_with` was first proposed, and Rust has evolved as a language since then. It's not clear the same soundness problems would matter as much now to Rust, or what mitigations would be acceptable, or how opsem differences matter, or if a changed type like clarfonthey proposes would be enough to carry it, or etc.

However, one would need to first absorb _why_ `replace_with` failed in 2016 before they could make it succeed in 2025. You'd need to identify which of the grounds it was effectively rejected on could now themselves be dismissed.

We can help you do so, but that's a nontrivial task, and I don't believe it's the sort of thing that will happen over a single PR of iterations, so I am closing this for now.

**Pull request closed without merging**